src/registry/http.ts

```diff
@@ -0,0 +1,439 @@
+      );
+    }
+
+    const t = await res.text();
+    try {
+      const response: {
+        access_token?: string;
+        expires_in: number;
+        repository: string;
+        token?: string;
+      } = JSON.parse(t);
+
+      console.debug(
+        `Authenticated with registry ${this.url.toString()} successfully, got token that expires in ${
+          response.expires_in
+        } seconds`,
+      );
+
+      if (!response.access_token && !response.token) {
+        console.error(
+          "Oauth response doesn't have access_token field, doing fallback to password_env, however this might mean that we will 401 later",
+        );
+      }
+
+      return {
+        authContext: ctx,
+        repository: response.repository ?? this.url.pathname,
+        accessToken: response.access_token ?? response.token ?? this.authBase64(),
+      };
+    } catch (err) {
+      console.error(
+        "Doing json response in authentication: ",
+        errorString(err),
+        t.slice(0, Math.min(t.length, 100)),
+        "status",
+        res.status,
+      );
+      throw err;
+    }
+  }
+
+  async authenticateBasic(ctx: AuthContext): Promise<HTTPContext> {
+    const res = await fetch(ctx.realm, {
+      headers: {
+        Authorization: "Basic " + this.authBase64(),
+      },
+    });
+
+    if (!res.ok) {
+      throw new Error(`couldn't authenticate with registry ${ctx.realm}: ${JSON.stringify(await res.json())}`);
+    }
+
+    return {
+      authContext: ctx,
+      accessToken: this.authBase64(),
+      repository: this.url.pathname.slice(1),
+    };
+  }
+
+  async manifestExists(namespace: string, tag: string): Promise<CheckManifestResponse | RegistryError> {
+    try {
+      const ctx = await this.authenticate();
+      const req = ctxIntoRequest(ctx, this.url, "HEAD", `${namespace}/manifests/${tag}`);
+      const res = await fetch(req);
+      if (!res.ok && res.status !== 404) {
+        console.warn(req.url, "->", res.status, "getting manifest:", await res.text());
+        return {
+          response: res,
+        };
+      }
+
+      console.log("->", req.url, res.status);
+      return {
+        exists: res.ok,
+        digest: res.headers.get("Docker-Content-Digest") as string,
+        size: +(res.headers.get("Content-Length") ?? "0"),
+        contentType: res.headers.get("Content-Type") ?? "",
+      };
+    } catch (err) {
+      console.error(`Error doing manifest exists with ${namespace} and ${tag}: ` + errorString(err));
+      return {
+        response: new InternalError(),
+      };
+    }
+  }
+
+  async getManifest(namespace: string, digest: string): Promise<GetManifestResponse | RegistryError> {
+    try {
+      const ctx = await this.authenticate();
+      const res = await fetch(ctxIntoRequest(ctx, this.url, "GET", `${namespace}/manifests/${digest}`));
+      if (!res.ok) {
+        return {
+          response: res,
+        };
+      }
```

Comment:
I'm getting this response here:

```
"{\"errors\":[{\"code\":\"MANIFEST_UNKNOWN\",\"message\":\"OCI index found, but Accept header does not support OCI indexes\"}]}\n"
```

Reply:
Thank you for trying it out, I updated it to have the header now.

Reply:
I tested it with Github and it seems to work now.